gpkit/tools/autosweep.py

```diff
@@ -0,0 +1,254 @@
+        if self.splitval:
+            raise ValueError("split already exists!")
+        if splitval <= self.bounds[0] or splitval >= self.bounds[1]:
+            raise ValueError("split value is at or outside bounds.")
+        self.splitval = splitval
+        self.splits = [BinarySweepTree([self.bounds[0], splitval],
+                                       [self.sols[0], splitsol],
+                                       self.sweptvar, self.costposy),
+                       BinarySweepTree([splitval, self.bounds[1]],
+                                       [splitsol, self.sols[1]],
+                                       self.sweptvar, self.costposy)]
+
+    def add_splitcost(self, splitval, splitlb, splitub):
+        "Adds a splitval, lower bound, and upper bound"
+        if self.splitval:
+            raise ValueError("split already exists!")
+        if splitval <= self.bounds[0] or splitval >= self.bounds[1]:
+            raise ValueError("split value is at or outside bounds.")
+        self.splitval = splitval
+        self.splitlb, self.splitub = splitlb, splitub
+
+    def posy_at(self, posy, value):
+        """Logspace interpolates between sols to get posynomial values.
+
+        No guarantees, just like a regular sweep.
+        """
+        if value < self.bounds[0] or value > self.bounds[1]:
+            raise ValueError("query value is outside bounds.")
+        bst = self.min_bst(value)
+        lo, hi = bst.bounds
+        loval, hival = [sol(posy) for sol in bst.sols]
+        lo, hi, loval, hival = log(map(mag, [lo, hi, loval, hival]))
+        interp = (hi-log(value))/float(hi-lo)
+        return exp(interp*loval + (1-interp)*hival)
+
+    def cost_at(self, _, value, bound=None):
+        "Logspace interpolates between split and costs. Guaranteed bounded."
+        if value < self.bounds[0] or value > self.bounds[1]:
+            raise ValueError("query value is outside bounds.")
+        bst = self.min_bst(value)
+        if bst.splitlb:
+            if bound:
+                if bound is "lb":
+                    splitcost = exp(bst.splitlb)
+                elif bound is "ub":
+                    splitcost = exp(bst.splitub)
+            else:
+                splitcost = exp((bst.splitlb + bst.splitub)/2)
+            if value <= bst.splitval:
+                lo, hi = bst.bounds[0], bst.splitval
+                loval, hival = bst.sols[0]["cost"], splitcost
+            else:
+                lo, hi = bst.splitval, bst.bounds[1]
+                loval, hival = splitcost, bst.sols[1]["cost"]
+        else:
+            lo, hi = bst.bounds
+            loval, hival = [sol["cost"] for sol in bst.sols]
+        lo, hi, loval, hival = log(map(mag, [lo, hi, loval, hival]))
+        interp = (hi-log(value))/float(hi-lo)
+        return exp(interp*loval + (1-interp)*hival)
+
+    def min_bst(self, value):
+        "Returns smallest spanning tree around value."
```

Comment:
you are using spanning tree here in a different way than the standard graph meaning, correct?

https://en.wikipedia.org/wiki/Spanning_tree